### PR TITLE
[SPARK-18518][ML] HasSolver supports override

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
1, make param support non-final with `finalFields` option
2, generate `HasSolver` with `finalFields = false`
3, override `solver` in LiR, GLR, and make MLPC inherit `HasSolver`

## How was this patch tested?
existing tests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
Jenkins, retest this please

**Comment by zhengruifeng** [comment]
cc @yanboliang

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala [hunk 1]
```diff
@@ -101,7 +100,7 @@ private[classification] trait MultilayerPerceptronParams extends PredictorParams
   final def getInitialWeights: Vector = $(initialWeights)
 
   setDefault(maxIter -> 100, tol -> 1e-6, blockSize -> 128,
-    solver -> MultilayerPerceptronClassifier.LBFGS, stepSize -> 0.03)
+    solver -> LBFGS, stepSize -> 0.03)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Why this change? is it for consistency? I usually like not importing class members unless it significantly improves readability since it slightly obscures the source, but, this is a side point

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 2]
```diff
@@ -421,6 +435,18 @@ object LinearRegression extends DefaultParamsReadable[LinearRegression] {
    */
   @Since("2.1.0")
   val MAX_FEATURES_FOR_NORMAL_SOLVER: Int = WeightedLeastSquares.MAX_NUM_FEATURES
+
+  /** String name for "auto". */
+  private[regression] val AUTO = "auto"
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: I am probably forgetting, but I thought we decided against introducing constants in a similar situation last month? not sure if it's the same context

Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: I modified here to follow the way in the companion object of `MultilayerPerceptronClassifier`. I don't know the discussion about whether to introduce constants ...

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala (hunk 1 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Just for consistency.

**Comment by zhengruifeng** [comment]
ping @yanboliang ?

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 3]
```diff
@@ -80,7 +80,7 @@ private[shared] object SharedParamsCodeGen {
       ParamDesc[String]("weightCol", "weight column name. If this is not set or empty, we treat " +
         "all instance weights as 1.0"),
       ParamDesc[String]("solver", "the solver algorithm for optimization. If this is not set or " +
-        "empty, default value is 'auto'", Some("\"auto\"")),
+        "empty, default value is 'auto'", finalFields = false),
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: Why remove default value? Can we still keep a default value for the most common case?

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 4]
```diff
@@ -97,6 +97,7 @@ private[shared] object SharedParamsCodeGen {
       doc: String,
       defaultValueStr: Option[String] = None,
       isValid: String = "",
+      finalFields: Boolean = true,
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: Can we combine ```finalFields``` and ```finalMethods``` as ```is FinalParam```? If the set/get methods are non-final, it makes sense to get a non-final param.

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 5]
```diff
@@ -167,6 +168,8 @@ private[shared] object SharedParamsCodeGen {
       "def"
     }
 
+    val finalStr = if (param.finalFields) "final " else ""
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: Remove blank from ```"final "```, please refer ```methodStr```.

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala (hunk 3 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: @yanboliang  Some alg like `GLR` do not support solver `auto` now.

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala (hunk 5 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: OK

**Comment by zhengruifeng** [comment]
Jenkins, retest this please

**Comment by zhengruifeng** [comment]
Jenkins, retest this please

**Comment by zhengruifeng** [comment]
Jenkins, retest this please

**Comment by zhengruifeng** [comment]
retest this please

File: python/pyspark/ml/classification.py [hunk 6]
```diff
@@ -1327,8 +1327,6 @@ class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol,
                       "remaining data in a partition then it is adjusted to the size of this " +
                       "data. Recommended size is between 10 and 1000, default is 128.",
                       typeConverter=TypeConverters.toInt)
-    solver = Param(Params._dummy(), "solver", "The solver algorithm for optimization. Supported " +
```
Comment (yanboliang) [inline python/pyspark/ml/classification.py]: We still need to print out supported options for different algorithms, so could we override param definition in Python as well as what we did in Scala?

**Review by yanboliang — COMMENTED** [review]
Only one comment, otherwise, LGTM. Thanks.

**Comment by zhengruifeng** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala [hunk 7]
```diff
@@ -75,17 +78,13 @@ private[classification] trait MultilayerPerceptronParams extends PredictorParams
    * Supported options: "gd" (minibatch gradient descent) or "l-bfgs".
    * Default: "l-bfgs"
    *
-   * @group expertParam
+   * @group param
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: Here should be ```expertParam```.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 8]
```diff
@@ -53,7 +53,23 @@ import org.apache.spark.storage.StorageLevel
...
+  import LinearRegression._
+
+  /**
+   * The solver algorithm for optimization.
+   * Supported options: "l-bfgs", "normal" and "auto".
+   * Default: "auto"
+   *
+   * @group expertParam
+   */
+  @Since("2.3.0")
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: ```expertParam``` -> ```param```

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 2 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: In MLlib convention, ```AUTO``` should be ```Auto```, but we can resolve it in #17995 .

**Review by yanboliang — APPROVED** [review]
LGTM, merged into master. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala [hunk 9]
```diff
@@ -143,7 +143,18 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
...
 
-  import GeneralizedLinearRegression._
+  /**
+   * The solver algorithm for optimization.
+   * Supported options: "irls" (iteratively reweighted least squares).
+   * Default: "irls"
+   *
+   * @group param
+   */
+  @Since("2.3.0")
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: 2.3.0 -> 2.0.0, please fix it in #17995 .

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 10]
```diff
@@ -53,7 +53,23 @@ import org.apache.spark.storage.StorageLevel
...
+  import LinearRegression._
+
+  /**
+   * The solver algorithm for optimization.
+   * Supported options: "l-bfgs", "normal" and "auto".
+   * Default: "auto"
+   *
+   * @group param
+   */
+  @Since("2.3.0")
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: 2.3.0 -> 1.6.0
